### PR TITLE
Custom params and token refresh callback

### PR DESCRIPTION
**Pull request by jvatic**
ping @titanous

### CODE REVIEW

File: oauth2.go [hunk 1]
```diff
@@ -149,6 +151,22 @@ func (c *Config) PasswordCredentialsToken(ctx context.Context, username, passwor
 	})
 }
 
+func (c *Config) getParams(base url.Values) url.Values {
+	var params url.Values
+	if c.Params != nil {
+		params = *c.Params
+	} else {
+		params = url.Values{}
```
Comment (titanous) [inline oauth2.go]: Use `make` with a capacity hint.

File: oauth2.go [hunk 2]
```diff
@@ -149,6 +151,22 @@ func (c *Config) PasswordCredentialsToken(ctx context.Context, username, passwor
 	})
 }
 
+func (c *Config) getParams(base url.Values) url.Values {
+	var params url.Values
+	if c.Params != nil {
+		params = *c.Params
```
Comment (titanous) [inline oauth2.go]: This doesn't make a copy of `c.Params`, you have to copy each element in a map as the underlying data structure of a map contains a pointer not the actual data.

File: token.go [hunk 3]
```diff
@@ -139,5 +143,9 @@ func retrieveToken(ctx context.Context, c *Config, v url.Values) (*Token, error)
 	if err != nil {
 		return nil, err
 	}
-	return tokenFromInternal(tk), nil
+	token := tokenFromInternal(tk)
+	if notifierFunc, ok := ctx.Value(TokenRefreshNotifier).(TokenRefreshNotifierFunc); ok {
+		notifierFunc(token)
+	}
```
Comment (titanous) [inline token.go:149]: I'm unclear on what this does?

Comment (jvatic) [inline token.go:149]: This is to propagate the new token back to the lib consumer whenever it gets refreshed ([example usage](https://github.com/flynn/flynn/blob/installer-azure/installer/azure_cluster.go#L22)).

**Comment by jvatic** [comment]
@titanous comments addressed

File: oauth2.go [hunk 4]
```diff
@@ -149,6 +151,31 @@ func (c *Config) PasswordCredentialsToken(ctx context.Context, username, passwor
...
+		params = make(url.Values, len(base))
+	}
+	for key, values := range base {
+		if _, ok := params[key]; ok {
+			continue
+		}
+		params[key] = make([]string, 0, len(values))
+		for _, v := range values {
+			params[key] = append(params[key], v)
+		}
```
Comment (titanous) [inline oauth2.go]: These four lines look the same as above, make a `cp` function inline inside this method?

Comment (jvatic) [inline oauth2.go]: Done

File: oauth2.go [hunk 5]
```diff
@@ -44,6 +44,8 @@ type Config struct {
 
 	// Scope specifies optional requested permissions.
 	Scopes []string
+
+	Params *url.Values
```
Comment (titanous) [inline oauth2.go]: No need for this to be a pointer

File: oauth2.go [hunk 6]
```diff
@@ -149,6 +151,32 @@ func (c *Config) PasswordCredentialsToken(ctx context.Context, username, passwor
 	})
 }
 
+func (c *Config) getParams(base url.Values) url.Values {
+	copyValues := func(values []string) []string {
+		newValues := make([]string, 0, len(values))
+		for _, v := range values {
+			newValues = append(newValues, v)
+		}
```
Comment (titanous) [inline oauth2.go]: drop the capacity and set the length, then just do `copy(newValues, v)`

**Comment by jvatic** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
LGTM

**Pull request merged**